### PR TITLE
feat: Preview icon themes

### PR DESCRIPTION
**Pull request by joshuamegnauth54**
(Maybe) closes: #208 

icon_picker

Edit: Actually it may be a little different than #208 but it still seems pretty useful.

### CODE REVIEW

**Comment by mmstick** [comment]
Requires design approval. @pop-os/ux

**Comment by git-f0x** [comment]
Personally, it looks a bit busy when always visible like that.
Maybe have icon theme selection in a separate window (similar to the Add Applet window), and in the place of the current drop-down for icon selection, maybe have a clickable name of the current theme, which opens the icon theme selection window.

**Comment by maria-komarova** [comment]
The features have been added before the design was finalized. The design is described here now: [#235](https://github.com/pop-os/cosmic-settings/issues/235). Please let me know if there are any questions.

**Comment by joshuamegnauth54** [comment]
Thank you for your comments, everyone! I'll try to implement these changes.

**Comment by maria-komarova** [comment]
Thanks for working on this!

**Comment by joshuamegnauth54** [comment]
theme_prev01
theme_prev02

I tested it with a few random themes that I placed into a fake `XDG_DATA_DIRS` to make sure it works.

File: cosmic-settings/src/pages/desktop/appearance.rs [hunk 1]
```diff
@@ -439,6 +446,45 @@ impl Page {
             .map(crate::pages::Message::Appearance)
     }
 
+    fn experimental_context_view(&self) -> Element<'_, crate::pages::Message> {
+        let active = self.icon_theme_active;
+        let theme = cosmic::theme::active();
+        let theme = theme.cosmic();
+        cosmic::iced::widget::column![
+            // Export theme choice to GNOME
```
Comment (mmstick) [inline cosmic-settings/src/pages/desktop/appearance.rs:454]: This option will apply to other toolkits as well. Not just GTK/GNOME.

File: cosmic-settings/src/pages/desktop/appearance.rs [hunk 2]
```diff
@@ -1571,3 +1631,102 @@ async fn set_gnome_icon_theme(theme: String) {
...
+    handles: &[icon::Handle],
+    id: usize,
+    selected: bool,
+) -> Element<'static, Message> {
+    let theme = cosmic::theme::active();
+    let theme = theme.cosmic();
+    cosmic::iced::widget::column![
+        button(
+            cosmic::iced::widget::column![
+                row::Row::new()
```
Comment (mmstick) [inline cosmic-settings/src/pages/desktop/appearance.rs:1679]: use `cosmic::widget::row()` instead.

File: cosmic-settings/src/pages/desktop/appearance.rs [hunk 3]
```diff
@@ -1571,3 +1631,102 @@ async fn set_gnome_icon_theme(theme: String) {
...
+/// Button with a preview of the icon theme.
+fn icon_theme_button(
+    name: &str,
+    handles: &[icon::Handle],
+    id: usize,
+    selected: bool,
+) -> Element<'static, Message> {
+    let theme = cosmic::theme::active();
+    let theme = theme.cosmic();
+    cosmic::iced::widget::column![
```
Comment (mmstick) [inline cosmic-settings/src/pages/desktop/appearance.rs:1676]: Use `cosmic::widget::column()` instead.

File: cosmic-settings/src/pages/desktop/appearance.rs [hunk 4]
```diff
@@ -829,6 +876,13 @@ impl Page {
                 }
                 Command::none()
             }
+            Message::ExperimentalContextDrawer => {
+                self.context_view = Some(ContextView::Experimental);
+
+                cosmic::command::message(crate::app::Message::OpenContextDrawer(
+                    fl!("experimental").into(),
```
Comment (mmstick) [inline cosmic-settings/src/pages/desktop/appearance.rs:883]: From the designs, this should be an empty string

**Comment by maria-komarova** [comment]
Looks great!

Just a few minor things base on the screenshot above (sorry if those were already fixed):
- The copy on the list item on the main content area should say "Experimental settings".
- There should be no header in the Context Drawer or if there is one, then is should say "Experimental Settings".
- The symbolic icon at the end of the list item on the main content area should be "go-next-symbolic" (actions icons).
- "Icon theme" inside Context drawer should be using "Header" text style and it should be spaced from the following thumbnails less (should be @space_xxs). The spacing between "Icon theme" and thumbnails should be less than the spacing between the list item above and "Icon theme" to maintain content hierarchy. 
- We might also want to cut the theme name after two lines. I know there is a text wrapping issue, it's been mentioned in other places. But still better have consistent spacing between rows of thumbnails. We can cut the long names with "...". 

Are icon theme thumbnails using image buttons? Or the same widget wallpapers do? It looks like it has a different corner radius and no additional selection indicator in the left bottom corner.

**Comment by maria-komarova** [comment]
Looks great!

Just a few minor things base on the screenshot above (sorry if those were already fixed):

- The copy on the list item on the main content area should say "Experimental settings".
- There should be no header in the Context Drawer or if there is one, then is should say "Experimental Settings".
- The symbolic icon at the end of the list item on the main content area should be "go-next-symbolic" (actions icons).
- "Icon theme" inside Context drawer should be using "Header" text style and it should be spaced from the following thumbnails less (should be @space_xxs). The spacing between "Icon theme" and thumbnails should be less than the spacing between the list item above and "Icon theme" to maintain content hierarchy.
- We might also want to cut the theme name after two lines. I know there is a text wrapping issue, it's been mentioned in other places. But still better have consistent spacing between rows of thumbnails. We can cut the long names with "...".

Are icon theme thumbnails using image buttons? Or the same widget wallpapers do? It looks like it has a different corner radius and no additional selection indicator in the left bottom corner.

**Comment by joshuamegnauth54** [comment]
menu
context

Here is an update to check if I'm on the right track.

To do tomorrow:
* Header text style
* Shortening long icon titles
* I think I have to pad the buttons a bit

**Comment by TheESN** [comment]
BTW It seems like this screen utilizes the smallest icon size variations from themes. Can you make it use bigger variations?

File: cosmic-settings/src/pages/desktop/appearance.rs [hunk 5]
```diff
@@ -1571,3 +1631,131 @@ async fn set_gnome_icon_theme(theme: String) {
...
+    selected: bool,
+) -> Element<'static, Message> {
+    let theme = cosmic::theme::active();
+    let theme = theme.cosmic();
+    // let image_style = cosmic::theme::Button::Image;
+    let background = theme.palette.neutral_4.into();
+
+    cosmic::widget::column()
+        .push(
+            button::Button::new_image(
```
Comment (mmstick) [inline cosmic-settings/src/pages/desktop/appearance.rs:1681]: This should be `cosmic::widget::button::image()`

File: cosmic-settings/src/pages/desktop/appearance.rs [hunk 6]
```diff
@@ -1571,3 +1631,131 @@ async fn set_gnome_icon_theme(theme: String) {
...
+                cosmic::iced::widget::column![
```
Comment (mmstick) [inline cosmic-settings/src/pages/desktop/appearance.rs:1682]: `cosmic::widget::column()`, `cosmic::widget::column::with_capacity()`, or `cosmic::widget::column::with_children()`.

File: cosmic-settings/src/pages/desktop/appearance.rs [hunk 7]
```diff
@@ -1571,3 +1631,131 @@ async fn set_gnome_icon_theme(theme: String) {
...
+                    cosmic::widget::row::row()
```
Comment (mmstick) [inline cosmic-settings/src/pages/desktop/appearance.rs:1683]: `cosmic::widget::row()`

File: cosmic-settings/src/pages/desktop/appearance.rs [hunk 8]
```diff
@@ -1571,3 +1631,131 @@ async fn set_gnome_icon_theme(theme: String) {
...
+                        .extend(
+                            handles
+                                .iter()
+                                .take(ICON_PREV_ROW)
+                                .cloned()
+                                .map(|handle| handle.icon().size(ICON_PREV_SIZE))
+                        )
+                        .spacing(theme.space_xs()),
+                    row::Row::new()
```
Comment (mmstick) [inline cosmic-settings/src/pages/desktop/appearance.rs:1692]: `cosmic::widget::row()`

File: cosmic-settings/src/pages/desktop/appearance.rs [hunk 9]
```diff
@@ -439,6 +446,48 @@ impl Page {
...
+            // Export theme choice
+            settings::view_section("").add(
+                settings::item::builder(fl!("enable-export"))
+                    .description(fl!("enable-export", "desc"))
+                    .toggler(self.tk.apply_theme_global, Message::ApplyThemeGlobal)
+            ),
+            // Icon theme previews
+            // cosmic::iced::widget::column![text(&*ICON_THEME), text(&*ICON_THEME_DESC).size(10)]
+            //     .spacing(2),
+            cosmic::iced::widget::column![
```
Comment (mmstick) [inline cosmic-settings/src/pages/desktop/appearance.rs:463]: `cosmic::widget::column()`, `cosmic::widget::column::with_capacity()`, or `cosmic::widget::column::with_children()`

**Review by mmstick — CHANGES_REQUESTED** [review]

**Comment by joshuamegnauth54** [comment]
Hello fellow cosmonauts!

Changes for today:
* Truncate long theme names
* Use [heading](https://pop-os.github.io/libcosmic/cosmic/widget/text/fn.heading.html) for "Icon theme"
* I padded the buttons a bit to fit the design better

new01

@TheESN I'm using `32` for the size since it seems like a good default, but I agree that some flexibility will be nice! Maybe a future patch can add a slider to increase or decrease the size?

I can also change the default for now too. I'm aesthetically challenged, so I'm not sure which size is best.

## Icon size 48
size_48

## Icon size 64
size_64

## Icon size 128
:joy_cat: 
size_128

32 (current) and 48 both look pretty nice.

**Comment by TheESN** [comment]
> Hello fellow cosmonauts!
> 
> Changes for today:
> 
>     * Truncate long theme names
> 
>     * Use [heading](https://pop-os.github.io/libcosmic/cosmic/widget/text/fn.heading.html) for "Icon theme"
> 
>     * I padded the buttons a bit to fit the design better
> 
> 
> new01
> 
> @TheESN I'm using `32` for the size since it seems like a good default, but I agree that some flexibility will be nice! Maybe a future patch can add a slider to increase or decrease the size?
> 
> I can also change the default for now too. I'm aesthetically challenged, so I'm not sure which size is best.
> ## Icon size 48
> 
> size_48
> ## Icon size 64
> 
> size_64
> ## Icon size 128
> 
> 😹 size_128
> 
> 32 (current) and 48 both look pretty nice.

I didn't mean the actual size, but their style. Icon themes usually specify several icons for different sizes, so smaller icons have less details, and bigger ones have more details. 

From screenshots that you sent I see that it uses the smallest size of icons and just scales them without changing to more detailed variations.

For reference that's how 32x32 Pop folder should look like
изображение

And that's how 48x48 should look like
изображение

The one that your preview is using is actually 16x16
изображение

**Comment by joshuamegnauth54** [comment]
> I didn't mean the actual size, but their style. Icon themes usually specify several icons for different sizes, so smaller icons have less details, and bigger ones have more details.
> 
> From screenshots that you sent I see that it uses the smallest size of icons and just scales them without changing to more detailed variations.

Alright. I think I fixed it now:

icon_size_fix

It turns out that I was evaluating the sizes too late. I updated the code to check for a few different sizes before falling back to the default too.

The situation for `La Capitaine` and `Luv` is different. I'm unsure if the problem rests with how I manually installed the themes or with `libcosmic`. Basically, the theme name and the directory have to match. `La Capitaine`'s `index.theme` lists the name as `Name=La Capitaine` but the directory is `la-capitaine`. I renamed the directory to match `index.theme` which worked.

The [spec](https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html) states:
> Each theme is stored as subdirectories of the base directories. The internal name of the theme is the name of the subdirectory, although the user-visible name as specified by the theme may be different. Hence, theme names are case sensitive, and are limited to ASCII characters. Theme names may also not contain comma or space.

The spec also lists an example of a theme titled `Birch` which is located in a directory called `birch`. I _think_ my issue with `La Capitaine` and `Luv` may be a bug?

**Comment by TheESN** [comment]
Looks definitely better! Though now I am wondering why adwaita is missing settings icon in preview

**Comment by joshuamegnauth54** [comment]
symbolic_fallback

Adwaita doesn't have `preferences-system` but it does have a symbolic iteration of that icon. I updated the code to try the icons listed in the design or fallback to the symbolic variant if the requested icon isn't found.

Alsooo, Vibrancy-Dark was missing icons because I didn't copy over the themes it inherited to my test `XDG_DATA_DIR`. I fixed it to ensure my code still works.

**Comment by TheESN** [comment]
> Adwaita doesn't have preferences-system but it does have a symbolic iteration of that icon

Well now I am wondering how is this icon called then?

изображение

**Comment by joshuamegnauth54** [comment]
I couldn't find an icon that looks like that in `Adwaita`. It's possible that that's an app icon stored in `hicolors` instead of `Adwaita` itself which would allow non-`Adwaita` themes to display the correct icon. `hicolors` is the default fallback for every theme which means it's a useful place to store app icons.

**Comment by TheESN** [comment]
> I couldn't find an icon that looks like that in `Adwaita`. It's possible that that's an app icon stored in `hicolors` instead of `Adwaita` itself which would allow non-`Adwaita` themes to display the correct icon. `hicolors` is the default fallback for every theme which means it's a useful place to store app icons.

Well maybe it's GNOME-only icon then. Kinda weird that they decided to go this way, so symbolic fallback should be fine then

**Review by mmstick — APPROVED** [review]

**Comment by mmstick** [comment]
Is this ready for merging?

**Review by maria-komarova — APPROVED** [review]

**Comment by maria-komarova** [comment]
Looks ready to me.

**Pull request merged**